File: library/src/main/java/com/proxerme/library/api/FixRatingDetailsAdapter.java
package com.proxerme.library.api;

import com.proxerme.library.entitiy.info.RatingDetails;
import com.squareup.moshi.FromJson;
import com.squareup.moshi.JsonAdapter;
import com.squareup.moshi.Moshi;
import com.squareup.moshi.ToJson;

import java.io.IOException;

/**
 * TODO: Describe class
 *
 * @author Ruben Gees
 */
class FixRatingDetailsAdapter {

    private static final String INVALID_ARRAY = "[]";

    private final JsonAdapter<RatingDetails> internalAdapter;

    FixRatingDetailsAdapter() {
        internalAdapter = new Moshi.Builder().build().adapter(RatingDetails.class);
    }

    @FromJson
    RatingDetails fromJson(final String json) throws IOException {
        if (json.equals(INVALID_ARRAY)) {
            return new RatingDetails(0, 0, 0, 0, 0);
        } else {
            return internalAdapter.fromJson(json);
        }
    }

    @ToJson
    String toJson(final RatingDetails ratingDetails) {
        return internalAdapter.toJson(ratingDetails);
    }
}

File: library/src/main/kotlin/me/proxer/library/internal/adapter/FixRatingDetailsAdapter.kt
package me.proxer.library.internal.adapter

import com.squareup.moshi.FromJson
import com.squareup.moshi.JsonAdapter
import com.squareup.moshi.JsonReader
import com.squareup.moshi.JsonWriter
import com.squareup.moshi.ToJson
import me.proxer.library.entity.info.RatingDetails

/**
 * @author Ruben Gees
 */
internal class FixRatingDetailsAdapter {

    private companion object {
        private const val INVALID_ARRAY = "[]"
    }

    @FromJson
    fun fromJson(jsonReader: JsonReader, delegate: JsonAdapter<RatingDetails>): RatingDetails? {
        val json = jsonReader.nextString()

        return when {
            json.isBlank() || json == INVALID_ARRAY -> RatingDetails(0, 0, 0, 0, 0)
            else -> delegate.fromJson(json)
        }
    }

    @ToJson
    fun toJson(writer: JsonWriter, value: RatingDetails?, delegate: JsonAdapter<RatingDetails>) {
        delegate.toJson(writer, value)
    }
}
